### PR TITLE
improve studio reporting robustness

### PR DESCRIPTION
**Pull request by garypen**
fixes: #1903 

The main fix here is that we previously shutdown the spaceport for the studio
reporting if we got a schema or config reload. That meant studio reporting would
never resume until the router re-started.
    
This change removes the shutdown spaceport on telemetry functionality
and instead relies on the fact that when a receiver gets no more traffic
it will terminate. This means we don't need to actively shut down a
spaceport connection.
    
There are some extra changes in here which I implemented along the way
to figuring out the main cause:
  - reconnect for recycled connections (not strictly required, but improves connection robustness)
  - extra logging to clarify control flow

***
fixes: #1893 

Rather than allowing deadpool to choose the maximum size of a
connection pool based on the execution environment, fix this at a
reasonable size (128) for all execution environments.

### CODE REVIEW

File: apollo-router/src/plugins/telemetry/mod.rs [hunk 1]
```diff
@@ -163,11 +162,7 @@ fn setup_metrics_exporter<T: MetricsConfigurator>(
 
 impl Drop for Telemetry {
     fn drop(&mut self) {
-        if let Some(sender) = self.spaceport_shutdown.take() {
-            ::tracing::debug!("notifying spaceport to shut down");
-            let _ = sender.send(());
-        }
-
+        ::tracing::info!("dropping telemetry...");
```
Comment (garypen) [inline apollo-router/src/plugins/telemetry/mod.rs:165]: make this debug

File: apollo-router/src/plugins/telemetry/mod.rs [hunk 2]
```diff
@@ -483,20 +472,25 @@ impl Telemetry {
...
+        // We're now safe for shutdown.
         // Start spaceport
         if let Some(spaceport) = spaceport {
             tokio::spawn(async move {
-                if let Err(e) = spaceport.serve().await {
-                    match e.source() {
+                ::tracing::debug!("serving spaceport");
+                match spaceport.serve().await {
+                    Ok(v) => {
+                        ::tracing::info!("spaceport terminated normally: {:?}", v);
```
Comment (garypen) [inline apollo-router/src/plugins/telemetry/mod.rs:482]: debug?

File: apollo-router/src/spaceport/mod.rs [hunk 3]
```diff
@@ -114,8 +115,17 @@ impl Reporter {
...
+        Ok(Self { client, ep })
+    }
+
+    /// Try to re-connect a reporter.
+    ///
+    /// This can fail if:
+    ///  - the reporter can't connect to the address
+    pub(crate) async fn reconnect(&mut self) -> Result<(), ReporterError> {
+        self.client = ReporterClient::connect(self.ep.clone()).await?;
+        Ok(())
```
Comment (garypen) [inline apollo-router/src/spaceport/mod.rs:128]: Not sure if needed. Possibly useful...

File: apollo-router/src/plugins/telemetry/apollo_exporter.rs [hunk 4]
```diff
@@ -175,12 +194,14 @@ impl managed::Manager for ReporterManager {
...
         let url = self.endpoint.to_string();
-        Ok(Reporter::try_new(url).await?)
+        Reporter::try_new(url).await
     }
 
-    async fn recycle(&self, _r: &mut Reporter) -> managed::RecycleResult<Self::Error> {
-        Ok(())
+    async fn recycle(&self, r: &mut Reporter) -> managed::RecycleResult<Self::Error> {
+        tracing::debug!("recycling reporter: {:?}", r);
+        r.reconnect().await.map_err(RecycleError::Backend)
```
Comment (garypen) [inline apollo-router/src/plugins/telemetry/apollo_exporter.rs:204]: Not sure if needed. Possibly useful.

**Review by Geal — APPROVED** [review]

**Pull request merged**